# Экстрактивная суммаризация с BERT

## Описание задачи

Необходимо провести суммаризацию текста, применив методы экстрактивной суммаризации.

Экстрактивная суммаризация – это метод автоматического извлечения ключевых предложений из исходного текста для формирования краткого содержания. В отличие от абстрактивной суммаризации, где создаются новые формулировки, экстрактивный подход берет существующие фразы и упорядочивает их в логичной последовательности.

Что нужно сделать?
Использовать BERT-модели – например, KeyBERT для выделения ключевых фраз, https://human.spbstu.ru/userfiles/files/articles/2024/1/20-35.pdf

Результат:
* код .py, .ipynb
* выводы


## Импорты и настройки

In [1]:
import pandas as pd

In [2]:
# Отображение до 50 столбцов в таблицах
pd.set_option('display.max_columns', 50)

# Сброс кастомного отображения столбцов в таблицах
# pd.reset_option('display.max_columns')

## Обзор данных 

Для данной задачи будем использовать уже распарсенный JSON чата, который был сохранён в CSV-формате

In [3]:
df = pd.read_csv("../def_Klyusnik_A/output/processed_messages_1.csv")

In [4]:
df.head(25)

name                type    id_first      id   type.1  \
0   💬 Data Practicum Chat  private_supergroup  1379846874  266690  service   
1   💬 Data Practicum Chat  private_supergroup  1379846874  266691  service   
2   💬 Data Practicum Chat  private_supergroup  1379846874  266693  message   
3   💬 Data Practicum Chat  private_supergroup  1379846874  266694  message   
4   💬 Data Practicum Chat  private_supergroup  1379846874  266695  message   
5   💬 Data Practicum Chat  private_supergroup  1379846874  266696  message   
6   💬 Data Practicum Chat  private_supergroup  1379846874  266697  message   
7   💬 Data Practicum Chat  private_supergroup  1379846874  266698  message   
8   💬 Data Practicum Chat  private_supergroup  1379846874  266699  message   
9   💬 Data Practicum Chat  private_supergroup  1379846874  266700  message   
10  💬 Data Practicum Chat  private_supergroup  1379846874  266701  message   
11  💬 Data Practicum Chat  private_supergroup  1379846874  266702  message   
12  💬 Data Practicum Chat  private_supergroup  1379846874  266703  message   
13  💬 Data Practicum Chat  private_supergroup  1379846874  266705  message   
14  💬 Data Practicum Chat  private_supergroup  1379846874  266707  message   
15  💬 Data Practicum Chat  private_supergroup  1379846874  266708  message   
16  💬 Data Practicum Chat  private_supergroup  1379846874  266709  message   
17  💬 Data Practicum Chat  private_supergroup  1379846874  266710  message   
18  💬 Data Practicum Chat  private_supergroup  1379846874  266712  message   
19  💬 Data Practicum Chat  private_supergroup  1379846874  266713  message   
20  💬 Data Practicum Chat  private_supergroup  1379846874  266715  message   
21  💬 Data Practicum Chat  private_supergroup  1379846874  266717  message   
22  💬 Data Practicum Chat  private_supergroup  1379846874  266718  message   
23  💬 Data Practicum Chat  private_supergroup  1379846874  266720  message   
24  💬 Data Practicum Chat  private_supergroup  1379846874  266722  message   

                   date  date_unixtime            actor       actor_id  \
0   2025-02-01T22:40:19     1738438819        Елизавета  user200103497   
1   2025-02-02T21:59:34     1738522774  Agamet Agametov  user217302209   
2   2025-02-03T11:28:38     1738571318              NaN            NaN   
3   2025-02-03T11:52:20     1738572740              NaN            NaN   
4   2025-02-03T11:52:37     1738572757              NaN            NaN   
5   2025-02-03T11:55:09     1738572909              NaN            NaN   
6   2025-02-03T11:56:57     1738573017              NaN            NaN   
7   2025-02-03T11:57:17     1738573037              NaN            NaN   
8   2025-02-03T11:57:36     1738573056              NaN            NaN   
9   2025-02-03T11:58:06     1738573086              NaN            NaN   
10  2025-02-03T11:59:48     1738573188              NaN            NaN   
11  2025-02-03T11:59:53     1738573193              NaN            NaN   
12  2025-02-03T12:02:55     1738573375              NaN            NaN   
13  2025-02-03T12:03:41     1738573421              NaN            NaN   
14  2025-02-03T12:04:18     1738573458              NaN            NaN   
15  2025-02-03T12:04:32     1738573472              NaN            NaN   
16  2025-02-03T12:04:34     1738573474              NaN            NaN   
17  2025-02-03T12:05:52     1738573552              NaN            NaN   
18  2025-02-03T12:06:56     1738573616              NaN            NaN   
19  2025-02-03T12:10:16     1738573816              NaN            NaN   
20  2025-02-03T12:10:29     1738573829              NaN            NaN   
21  2025-02-03T12:10:36     1738573836              NaN            NaN   
22  2025-02-03T12:10:42     1738573842              NaN            NaN   
23  2025-02-03T12:12:19     1738573939              NaN            NaN   
24  2025-02-03T12:12:27     1738573947              NaN            NaN   

                action inviter  \
0   join_group_by_link   Group   

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 36 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   name                 1007 non-null   object 
 1   type                 1007 non-null   object 
 2   id_first             1007 non-null   int64  
 3   id                   1007 non-null   int64  
 4   type.1               1007 non-null   object 
 5   date                 1007 non-null   object 
 6   date_unixtime        1007 non-null   int64  
 7   actor                83 non-null     object 
 8   actor_id             83 non-null     object 
 9   action               83 non-null     object 
 10  inviter              81 non-null     object 
 11  text                 881 non-null    object 
 12  text_entities        1007 non-null   object 
 13  from                 924 non-null    object 
 14  from_id              924 non-null    object 
 15  reply_to_message_id  689 non-null    f

In [6]:
df.text.head(15)

0                                                   NaN
1                                                   NaN
2     ['Всем большой привет! Приглашаю на свой уютны...
3        А у тебя когда будет свой канал про аналитику?
4           Будешь туда голосовухи пятиминутные постить
5     Потому что сделаны так, будто устарели уже лет...
6                                             Подкаст?)
7                             Не, это не так раздражает
8     Нужны голосовуки с эээ, нууу, и 10 секундными ...
9     Звонили из ада, просили передать что ждут на м...
10                Ты хотел сказать "ждут мастер-класс"?
11    Их сделали, чтобы понравится человеку, котором...
12                      Благодарствую, ещё не проснулся
13    Боже теперь и я в этом рейтинге😆😆😆 и тебе спасибо
14                                          зря конечно
Name: text, dtype: object

In [7]:
df.text_entities.head(15)

0                                                    []
1                                                    []
2     [{'type': 'plain', 'text': 'Всем большой приве...
3     [{'type': 'plain', 'text': 'А у тебя когда буд...
4     [{'type': 'plain', 'text': 'Будешь туда голосо...
5     [{'type': 'plain', 'text': 'Потому что сделаны...
6              [{'type': 'plain', 'text': 'Подкаст?)'}]
7     [{'type': 'plain', 'text': 'Не, это не так раз...
8     [{'type': 'plain', 'text': 'Нужны голосовуки с...
9     [{'type': 'plain', 'text': 'Звонили из ада, пр...
10    [{'type': 'plain', 'text': 'Ты хотел сказать "...
11    [{'type': 'plain', 'text': 'Их сделали, чтобы ...
12    [{'type': 'plain', 'text': 'Благодарствую, ещё...
13    [{'type': 'plain', 'text': 'Боже теперь и я в ...
14           [{'type': 'plain', 'text': 'зря конечно'}]
Name: text_entities, dtype: object

### Вывод

В нашем распоряжении имеется выгрузка чата, которая содержит 1007 строк и 36 столбцов. Имеется множество пропусков в данных.

Ключевые признаки для нас: `text` и `text_entities`

## EDA